metadata/metadata.go

```diff
@@ -105,3 +147,41 @@ func safeCopy(destination map[string]string, key, val string) {
 		destination[key] = val
 	}
 }
+
+// copyWithNewPrefix copies key/value pairs from a source map to a destination map where the key matches the specified prefix.
+// If replacementPrefix is different from prefix, the prefix will be replaced while performing the copy.
+func copyWithNewPrefix(src, dest map[string]string, prefix, replacementPrefix string) {
+	for key, value := range src {
+		if strings.HasPrefix(key, prefix) {
+			newKey := key
+			if prefix != replacementPrefix {
+				newKey = strings.Replace(key, prefix, replacementPrefix, 1)
+			}
+			dest[newKey] = value
+		}
+	}
+}
+
+// CopyLabelsByPrefix copies all labels from a source object to a destination object where the key matches the specified prefix.
+// If replacementPrefix is different from prefix, the prefix will be replaced while performing the copy.
+func CopyLabelsByPrefix(src, dest v1.Object, prefix, replacementPrefix string) {
+	if src.GetLabels() == nil {
+		return
+	}
+	if dest.GetLabels() == nil {
+		dest.SetLabels(make(map[string]string))
+	}
+	copyWithNewPrefix(src.GetLabels(), dest.GetLabels(), prefix, replacementPrefix)
+}
+
+// CopyAnnotationsByPrefix copies all annotations from a source object to a destination object where the key matches the specified prefix.
+// If replacementPrefix is different from prefix, the prefix will be replaced while performing the copy.
+func CopyAnnotationsByPrefix(src, dest v1.Object, prefix, replacementPrefix string) {
+	if src.GetAnnotations() == nil {
```

Comment:
We should check if the objects are nil instead. Check GetLabelsWithPrefix.

Reply:
Updated. Please help to review it again. Thanks